packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -329,48 +347,106 @@ export class Pipeline extends cdk.Construct implements cpapi.IPipeline {
+    const actionRole = this.attachActionForCrossAccount(action);
+
+    // call the Action callback which eventually calls bind()
+    (action as any)._actionAttachedToPipeline(stage,
+        actionRole === undefined ? this.role : actionRole,
+        actionScope, actionRole);
+  }
+
+  private attachActionForCrossRegion(region?: string): void {
+    if (!region) {
       return;
     }
 
     // get the region the Pipeline itself is in
     const pipelineRegion = this.node.stack.requireRegion(
-      "You need to specify an explicit region when using CodePipeline's cross-region support");
+        "You need to specify an explicit region when using CodePipeline's cross-region support");
 
     // if we already have an ArtifactStore generated for this region, or it's the Pipeline's region, nothing to do
-    if (this.artifactStores[action.region] || action.region === pipelineRegion) {
+    if (this.artifactStores[region] || region === pipelineRegion) {
       return;
     }
 
-    let replicationBucketName = this.crossRegionReplicationBuckets[action.region];
+    let replicationBucketName = this.crossRegionReplicationBuckets[region];
     if (!replicationBucketName) {
       const pipelineAccount = this.node.stack.requireAccountId(
-        "You need to specify an explicit account when using CodePipeline's cross-region support");
+          "You need to specify an explicit account when using CodePipeline's cross-region support");
       const app = this.node.stack.parentApp();
       if (!app) {
         throw new Error(`Pipeline stack which uses cross region actions must be part of an application`);
       }
       const crossRegionScaffoldStack = new CrossRegionScaffoldStack(app, {
-        region: action.region,
+        region,
         account: pipelineAccount,
       });
-      this._crossRegionScaffoldStacks[action.region] = crossRegionScaffoldStack;
+      this._crossRegionScaffoldStacks[region] = crossRegionScaffoldStack;
       replicationBucketName = crossRegionScaffoldStack.replicationBucketName;
     }
 
-    const replicationBucket = s3.Bucket.import(this, 'CrossRegionCodePipelineReplicationBucket-' + action.region, {
+    const replicationBucket = s3.Bucket.import(this, 'CrossRegionCodePipelineReplicationBucket-' + region, {
       bucketName: replicationBucketName,
     });
     replicationBucket.grantReadWrite(this.role);
 
-    this.artifactStores[action.region] = {
+    this.artifactStores[region] = {
       Location: replicationBucket.bucketName,
       Type: 'S3',
     };
   }
 
+  private attachActionForCrossAccount(action: cpapi.Action): iam.IRole | undefined {
+    if (!action.resource) {
+      return action.role;
+    }
+
+    const pipelineStack = this.node.stack;
+    const resourceStack = action.resource.node.stack;
+    if (pipelineStack.env.account !== resourceStack.env.account) {
+      if (!this.artifactBucket.encryptionKey) {
+        throw new Error('The Pipeline is being used in a cross-account manner, ' +
```

Comment:
Is this physically true or policy we are setting?

Reply:
(For whatever "physically" means in this context...)

Reply:
It is required for cross-account Pipelines, yes.